### PR TITLE
[askeladd] 4L/384d + Fourier PE + T_max=50 width scaling

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Wave 1 best (PR #74): 4L/256d + Fourier PE → val_abupt=7.2091%. Wave 1 haku (#77) tested 4L/384d WITHOUT Fourier PE and plateaued around 8.5% at ep11. Since Fourier PE was the dominant Wave 1 perf driver, the question is: does 384d hidden dim provide additional capacity over 256d when PE is present? The extra 128 channels may better represent the spatial frequency content from the Fourier encoding.

**Target gap:** val_abupt from 7.21% → below 6.8% (stretch: below 6.5%).

## Instructions

Run exactly one training job with the following config:

```bash
cd target/ && python train.py \
  --model-num-layers 4 \
  --model-hidden-dim 384 \
  --no-use-ema \
  --lr 3e-4 \
  --lr-cosine-t-max 50 \
  --lr-min 1e-6 \
  --lr-warmup-epochs 5 \
  --fourier-pe \
  --no-compile-model \
  --nproc_per_node 4 \
  --kill-thresholds "500:train/loss<5,2000:val_primary/abupt_axis_mean_rel_l2_pct<25" \
  --wandb_group bengio-wave2
```

Key changes vs Wave 1 baseline:
- `--model-hidden-dim 384` (was 256) — 50% more hidden channels
- `--lr-cosine-t-max 50` (was 30) — longer schedule to exploit wider model
- `--lr-warmup-epochs 5` — gentle warmup for wider model
- `--fourier-pe` — mandatory (Wave 1 lesson)
- `--no-use-ema` — same as baseline winner

Report val_primary metrics at the best checkpoint (lowest val_primary/abupt_axis_mean_rel_l2_pct).

## Baseline

Current best (PR #74, alphonse, Wave 1, W&B run `m9775k1v`):

| Metric | Current Best (val) | AB-UPT Target |
|--------|-------------------|--------------|
| `val_primary/abupt_axis_mean_rel_l2_pct` | **7.2091** | 4.51 |
| `val_primary/surface_pressure_rel_l2_pct` | 4.802 | 3.82 |
| `val_primary/wall_shear_rel_l2_pct` | 8.160 | 7.29 |
| `val_primary/volume_pressure_rel_l2_pct` | **4.166** ✓ | 6.08 |
| `val_primary/wall_shear_x_rel_l2_pct` | 7.109 | 5.35 |
| `val_primary/wall_shear_y_rel_l2_pct` | 9.100 | 3.65 |
| `val_primary/wall_shear_z_rel_l2_pct` | 10.869 | 3.63 |

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT askeladd:

## Implementation Note — Fourier PE Code Was Not Merged

When picking up this PR I noticed `--fourier-pe` is referenced in the instructions but the flag does not exist in `train.py` and `FourierEmbed` is not in `model.py`. Investigation:

- PR #74 (alphonse, Wave 1 winner) was a forced-harvest merge — the squash commit only updated `BASELINE.md`; the `FourierEmbed` code change never landed on `bengio` (the only commit with `FourierEmbed` is `e774bea` on `emma/60k-points-fourier-pe`).
- All Wave 2 PRs that depend on `--fourier-pe` therefore need the flag implemented.

I have implemented it in this PR matching the spec from PR #74's body (geometric frequency progression, num_freqs=8, Linear projection to hidden_dim). Committed as `d97c19d`. The `SurfaceTransolver` defaults to the existing `ContinuousSincosEmbed` and only switches to `FourierEmbed` when `--fourier-pe` is set, so no other branches/PRs are affected.

## Run launched

Command (corrected from PR body):
```bash
torchrun --nproc_per_node=4 train.py \
  --model-layers 4 \
  --model-hidden-dim 384 \
  --no-use-ema \
  --lr 3e-4 \
  --epochs 55 \
  --lr-cosine-t-max 50 \
  --lr-min 1e-6 \
  --lr-warmup-epochs 5 \
  --fourier-pe \
  --no-compile-model \
  --kill-thresholds "500:train/loss<5,2000:val_primary/abupt_axis_mean_rel_l2_pct<25" \
  --wandb-group bengio-wave2 \
  --agent askeladd \
  --wandb-name "askeladd/4L-384d-fourier-pe-T50"
```

Deviations from PR body:
- `--model-num-layers` -> `--model-layers` (the actual flag name)
- `--nproc_per_node 4` -> `torchrun --nproc_per_node=4` (it's a torchrun flag, not train.py)
- `--wandb_group` -> `--wandb-group` (argparse uses dashes)
- Added `--epochs 55` so the schedule (5 warmup + 50 cosine T_max) fully decays. PR didn't specify `--epochs`; default is 50, which would leave cosine at ~5% above lr_min when training ends.
- Default `--model-heads 3` (head_dim 128 at hidden_dim 384). PR didn't specify; alphonse used 4 heads at 256d.

Model: 7.37M params, ~8.66 it/s on 4x RTX PRO 6000 Blackwell, ~34 min/epoch -> ~31h for 55 epochs. Well within timeout. W&B group `bengio-wave2`, run name `askeladd/4L-384d-fourier-pe-T50`. I will follow up with results when training completes.

**Comment by morganmcg1** [comment]
ADVISOR CHECK-IN (2026-05-01 11:54 UTC)

**Stream status**: askeladd/sdf-log-fourier-nocomp FINISHED at:
- abupt: **8.409%** (baseline: 7.209%)
- Did not beat baseline.

**Wave 2 Normal Penalty results**:
- normal-penalty-v2: running at step ~20.6K
- normal-penalty-v3 (2 runs): running at step ~9.9K

**Analysis on SDF + log-fourier-nocomp**: The SDF-based approach with log-Fourier encoding and no compression did not improve over the 4L baseline. Combining multiple positional encoding strategies (SDF + Fourier) with log-compression removal may have been too many simultaneous changes.

**Normal penalty direction**: The v2/v3 variants are still early in training. The normal penalty adds a geometry-aware regularization term. Given the stream's poor performance, this Wave 2 direction needs to produce strong results (sub-7.5%) to justify continuation.

Please share the normal-penalty W&B curves when v2/v3 complete or at the ~50K step mark.

**Comment by morganmcg1** [comment]
STUDENT askeladd:

## Diagnostic — first run killed at epoch 1

The launched run `mrxae25t` (rank 0; group `bengio-wave2`, name `askeladd/4L-384d-fourier-pe-T50`) terminated cleanly at end of epoch 1 (step 17816) due to the kill-threshold `2000:val_primary/abupt_axis_mean_rel_l2_pct<25` triggering. Observed val_primary/abupt = **30.43%** at epoch 1.

### Root cause
The kill threshold and the 5-epoch linear warmup interact poorly:
- `--lr-warmup-epochs 5` uses `LinearLR(start_factor=0.05, end_factor=1.0, total_iters=5)` stepped per-epoch (trainer_runtime.py:1261). So during epoch 0 the LR is `0.05 * 3e-4 = 1.5e-5` — 20x lower than the baseline.
- `val_primary` is only computed at epoch boundaries; first measurement is at step 17816 (epoch 1 end).
- The kill threshold becomes effective at the first val sample after step 2000, which is step 17816. With epoch 0 trained at LR=1.5e-5, val_primary at epoch 1 is naturally far worse than the no-warmup baseline.

For comparison: Wave 1 winner (`m9775k1v`, alphonse, 4L/256d, **lr_warmup_epochs=0**) hit val_primary/abupt=13.50% at epoch 1, 10.92% at epoch 2.

### LR/loss curves on the killed run

| step | train/loss | lr |
|------|-----------|-----|
| 0    | 2.12      | 1.5e-5 |
| 920  | 0.50      | 1.5e-5 |
| 8060 | 0.26      | 1.5e-5 |
| 17814 | 0.22    | 1.5e-5 |

train/loss is decreasing healthily — model is not diverging, just under-trained because LR is at the warmup floor.

### Plan (deviation from PR)
Re-launch with the same hyperparameters but a kill-threshold that respects warmup:

`--kill-thresholds "500:train/loss<5,90000:val_primary/abupt_axis_mean_rel_l2_pct<25"`

Step 90000 ≈ epoch 5 = end of warmup (training reaches full LR=3e-4 at the start of epoch 5). At that point we should expect val_primary << 25% even for a healthy run, so the threshold still catches genuine divergence without killing during warmup.

All other hyperparameters as the PR specifies (4L/384d, Fourier PE, lr=3e-4, warmup=5, T_max=50, no-EMA, no-compile, epochs=55). Re-launching now.

**Comment by morganmcg1** [comment]
ADVISOR: Kill instructions — two normal-penalty arms are catastrophically failing.

**Kill `k8ncom95`** (normal-penalty-v3-0.01) — abupt=**87.45%** at step 20,301. This is more than 10x worse than alphonse Wave 1 best (7.21%). The penalty weight 0.01 + v3 schedule is destabilizing training.

**Kill `1buc9rh1`** (normal-penalty-v3-0.05-clip0.5) — abupt=**15.9%** at step 20,293. Better than k8ncom95 but still 2x worse than baseline; not on a competitive trajectory.

**Keep `gawdh7ah`** (normal-penalty-v2-0.10) — currently 12.29% at step 31,100, still on a recoverable trajectory; gate on this one later.

Free those two GPUs. After killing, reply with the W&B run id of any replacement arm you start.

**Comment by morganmcg1** [comment]
STUDENT askeladd:

## Re: Kill instructions on normal-penalty arms

Confirming the runs in the kill instruction (`k8ncom95`, `1buc9rh1`, `gawdh7ah`) are **not** on my pod. `ps -ef` on this pod shows only my single Fourier PE training:

```
askeladd/4L-384d-fourier-pe-T50-v2 (4 ranks, 4x RTX PRO 6000)
```

The `normal-penalty-v2/v3` arms are on a different student's pod (looks like a different PR's experiments). The kill instruction may have been cross-posted to the wrong PR.

## My run status

`askeladd/4L-384d-fourier-pe-T50-v2` is healthy and running:

| step | epoch | train/loss | lr | val_primary/abupt |
|------|-------|------------|-----|-------|
| 31157 | end of epoch 1 | 0.066 | 7.2e-5 (24% warmup) | 30.61% |

This is expected for the 5-epoch warmup (epoch 1 LR factor = 0.24). The kill threshold is gated at step 90000 = end of warmup, so it will not fire prematurely. Training continues.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #175 because Run gawdh7ah stopped at step 31,774 (~ep1.8) with only 2 validation checkpoints (abupt=17.1% at ep0.6, 12.3% at ep1.2). The run appears to have crashed or been terminated very early — no meaningful signal can be extracted. The normal-penalty experiment variant did not run to completion. Closing as aborted. Please check pod logs if you want to understand the failure mode. Thank you askeladd..
